atp/client.go

```diff
@@ -401,7 +444,15 @@ func (c *client) executeReadLoop(cborReader *cbor.Decoder) {
 				runtimeMessage.MessageID)
 		}
 		c.mutex.Lock()
-		if len(c.runningStepResultChannels) == 0 {
+		remainingSteps := 0
+		for _, resultEntry := range c.runningStepResultEntries {
+			// The result is the reliable way to determine if it's done. There is a fraction of
+			// time when the entry is still in the map, but it's done.
+			if resultEntry.result == nil {
+				remainingSteps++
+			}
+		}
+		if remainingSteps == 0 {
 			c.mutex.Unlock()
 			return // Done
 		}
```

Comment:
I think you can reduce the size of this critical section to just the loop:  `remainingSteps` is a local variable, so you don't need the mutex to access it.  This means that you can do the initialization before entering the critical section, pull the `Unlock` from line 459 up to before the `if` statement, and remove the unlock duplication:
```go
		remainingSteps := 0
		c.mutex.Lock()
		for _, resultEntry := range c.runningStepResultEntries {
			if resultEntry.result == nil {
				remainingSteps++
			}
		}
		c.mutex.Unlock()
		if remainingSteps == 0 {
			return
		}
```
Also, with respect to the code comments....  The one on the return statement is superfluous.  And, I would be inclined to rephrase the one in the loop in the negative:  that is, we're looking for things which aren't done, and those are existing entries whose values are `nil`.  (I wouldn't bother explaining that there are two other states -- entries which have been removed and entries which are not `nil`.)

By the way, you might want to consider short-circuiting the loop:  you don't need to count all the entries, you only need to know if there is _any_ one which is `nil` -- once you find it, set a flag, and break out of the loop.  (That is, `remainingSteps` should be a boolean, not a count.)  E.g.:
```go
		remainingSteps := false
		c.mutex.Lock()
		for _, resultEntry := range c.runningStepResultEntries {
			if resultEntry.result == nil {
				remainingSteps = true
				break
			}
		}
		c.mutex.Unlock()
		if !remainingSteps {
			return
		}
```